### ISSUE TITLE
package/channel clash?

### ISSUE THREAD

**Issue by dhirschfeld** [conda-forge/python-simplegeneric-feedstock#1]
Having both the default conda repos and conda-forge in your .condarc will break conda because it doesn't recognise this as being the same package as simplegeneric in the default repos.

Is there any reason to duplicate packages which are already in the defaults? If so, wouldn't it be best to keep the same name so conda recognises it as the same package?

conda/conda#2773

**Comment by kalefranz** [comment]
Fairly sure this is a problem with conda 4.1.x we have not yet fixed by 4.1.2.  We're working on it.

**Comment by mcg1969** [comment]
This is not a conda problem. This is a problem caused by Conda-Forge's choice to name this differently than defaults. There is simply no way to prevent people from installing `simplegeneric` and `python-simplegeneric` on top of each other. Conda-forge needs to revert the name of this package back to `simplegeneric`.

**Comment by kalefranz** [comment]
OK conda-forge team, ecosystem namespaces are coming.  @mcg1969 is working on it.

Until then, I'm satisfied this explanation closes:

* https://github.com/conda/conda/issues/2673
* https://github.com/conda/conda/issues/2773

**Comment by ocefpaf** [comment]
@dhirschfeld and @kalefranz the story behind the rename to `python-simplegeneric` is in https://github.com/conda-forge/staged-recipes/pull/696

>  If so, wouldn't it be best to keep the same name so conda recognises it as the same package?

There is no way right now. Ideally we should try our best to use the same names and avoid conflict. However, as `conda-forge` grows faster than the pace defaults grows, we have the need to disambiguate the names faster too.

@dhirschfeld note that the namespaces mentioned by @kalefranz will help a lot here as we will be able to host to packages named `simplegeneric` (the `java` and the `python`) without conflict.

PS: We will need to take action here once the namespace feature gets implemented Do we rename this feedstock?

**Comment by mcg1969** [comment]
Yes, @ocefpaf, there *is* a way to do it *right now*. It's to keep the same name for the Python package, and to append a prefixe to the Java package.

I understand that favoring Python is undesirable in the long term, but _right now_ this is causing existing users difficulty. It's absurd to hold so tightly to some aesthetic ideal that we would be causing users issues like this.

**Comment by mcg1969** [comment]
A cursory check reveals we are going to have this problem with `drmaa`, `decorator`, `pathlib2` as well..

**Comment by dhirschfeld** [comment]
Why is does this package even exist at present? Everyone can get `simplegeneric` already from defaults so there is no need for this feedstock. This is especially true in light of the fact that the renaming breaks conda.

Why don't we remove this feedstock (or at least all currently existing packages) until such time as the namespace ability will allow this to not break conda?

**Comment by kalefranz** [comment]
FWIW, once we have ecosystem namespacing worked out, `defaults` will defer to the ecosystem's primary, flat, package distribution namespace for the canonical name of each package within an ecosystem.  Thus, in general, python package names should be as they are on PyPI, node package names as they are on npmjs.org, ruby package names as they are on rubygems.org, etc etc

**Comment by kalefranz** [comment]
Also reported at https://github.com/conda/conda/issues/2785, moved to #2

**Comment by msarahan** [comment]
@ocefpaf 

> There is no way right now. Ideally we should try our best to use the same names and avoid conflict. However, as conda-forge grows faster than the pace defaults grows, we have the need to disambiguate the names faster too.

conda-forge is growing faster, but we should respect the precedent in defaults where it exists.  This is just practical - don't break things.  The other side of this, in my opinion, is that defaults should respect precedents in conda-forge, where they exist.

@dhirschfeld 
> Why is does this package even exist at present? Everyone can get simplegeneric already from defaults so there is no need for this feedstock. This is especially true in light of the fact that the renaming breaks conda.

One critical detail is that packages on conda-forge are generally different recipes than defaults.  Sometimes newer versions are available on conda-forge, and sometimes fixes are available on conda-forge (looking at the GDAL world).  Just because something is already in defaults doesn't mean conda-forge should stay away.  However, it does mean conda-forge should take special precaution to prevent confusion and breakage.

**Comment by ocefpaf** [comment]
> conda-forge is growing faster, but we should respect the precedent in defaults where it exists.

I agree. Even though I am a big fan of the `python-` prefix we should align `defaults` with `conda-forge`. I will open issues on the packages that are current name differently to keep track of that.

**Comment by jakirkham** [comment]
So, I thought we initially agreed to do metapackages that included the original name from `defaults` to mitigate this exact sort of problem. However, this would only remain temporarily. Is that now also out? It seems this proposal could solve the problem seen here. Regardless, I don't really want to sink into this discussion.

@mcg1969, could you please open an issue on `conda` or `conda-build` as appropriate explaining how you think namespacing will work? My understanding of this plan is pretty fuzzy at this point and I imagine part of the push for prefixing has been that how this works is completely clear. Some explanation about this proposal may ease concerns. Plus I think there are many of us that would like to ask questions. Unfortunately now there is no home for them. So it will eventually have to be done over a range of partially related PRs and issues. This will ultimately make it harder to figure out what we decided and why.

**Comment by mcg1969** [comment]
> So, I thought we initially agreed to do metapackages that included the original name from defaults to mitigate this exact sort of problem.

Who is "we"?

Yes, we'll open up an issue to describe the namespace plan. We have been swamped lately. But it's a high priority for us so I'll get to it soon.

**Comment by jakirkham** [comment]
> Who is "we"?

The people at the last conda-forge meeting. We had agreed to prefix, which is why there is this conflict now.

 > Yes, we'll open up an issue to describe the namespace plan. We have been swamped lately. But it's a high priority for us so I'll get to it soon.

👍

No worries. Just make sure to let us know when you do.
